extensions/amp-subscriptions/0.1/viewer-tracker.js

```diff
@@ -0,0 +1,126 @@
+/**
+ * Copyright 2018 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {Services} from '../../../src/services';
+import {cancellation} from '../../../src/error';
+import {dev} from '../../../src/log';
+import {listenOnce} from '../../../src/event-helper';
+
+const TAG = 'local-viewer';
+
+export class ViewerTracker {
+  constructor(ampdoc) {
+    this.ampdoc_ = ampdoc;
+
+    /** @private @const {!../../../src/service/viewer-impl.Viewer} */
+    this.viewer_ = Services.viewerForDoc(ampdoc);
+
+    /** @private {?Promise} */
+    this.reportViewPromise_ = null;
+
+    /** @const @private {!../../../src/service/timer-impl.Timer} */
+    this.timer_ = Services.timerFor(ampdoc.win);
+
+    /** @private @const {!../../../src/service/viewport/viewport-impl.Viewport} */
+    this.viewport_ = Services.viewportForDoc(ampdoc);
+  }
+
+  /**
+   * @param {time} timeToView
+   * @returns {Promise}
+   */
+  scheduleView(timeToView) {
+    this.reportViewPromise_ = null;
+    return this.ampdoc_.whenReady().then(() => {
+      if (this.viewer_.isVisible()) {
+        return this.reportWhenViewed_(timeToView);
+      }
+      this.viewer_.onVisibilityChanged(() => {
+        if (this.viewer_.isVisible()) {
+          return this.reportWhenViewed_(timeToView);
```

Comment:
No point in this `return`. Which makes me thing that maybe this API is a bit broken? You are relying on this response to return the promise when "view" has happened, right?

My guess is that you need to do something like this:
```
return new Promise(resolve => {
  this.viewer_.onVisibilityChanged(() => {
    if (this.viewer_.isVisible()) {
      resolve();
    }
  });
}).then(() => this.reportWhenViewed_(...));
```

or similar.

Reply:
Done something similar